### ISSUE TITLE
[JENKINS-58290] WebSocket / OkHttp thread leak from BourneShellScript + ContainerExecDecorator

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2221]
<p>The <tt>durable-task</tt> plugin runs a wrapper process which redirects the user process' stdout/err to a file and sends its exit code to another file. Thus there is no need for the agent JVM to hold onto a process handle for the wrapper; it should be fork-and-forget. In fact the <tt>Proc</tt> is discarded.</p>

<p>Unfortunately, the current implementation in <tt>BourneShellScript</tt> does not actually allow the <tt>Proc</tt> to exit until the user process also exits. On a regular agent this does not matter much. But when you run <tt>sh</tt> steps inside <tt>container</tt> on a Kubernetes agent, <tt>ContainerExecDecorator</tt> and <tt>ContainerExecProc</tt> actually keep a WebSocket open for the duration of the launched process. This consumes resources on the Kubernetes API server; it is possible to run out of connections. It also consumes three master-side Java threads per <tt>sh</tt>, like</p>

<pre>
"OkHttp http://…/..." #361 prio=5 os_prio=0 tid=… nid=… runnable […]
   java.lang.Thread.State: RUNNABLE
at java.net.SocketInputStream.socketRead0(Native Method)
at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
at java.net.SocketInputStream.read(SocketInputStream.java:171)
at java.net.SocketInputStream.read(SocketInputStream.java:141)
at okio.Okio$2.read(Okio.java:140)
at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
at okio.RealBufferedSource.request(RealBufferedSource.java:68)
at okio.RealBufferedSource.require(RealBufferedSource.java:61)
at okio.RealBufferedSource.readByte(RealBufferedSource.java:74)
at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.java:117)
at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:101)
at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:274)
at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:214)
at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)

"OkHttp WebSocket http://…/..." #359 prio=5 os_prio=0 tid=… nid=… waiting on condition […]
   java.lang.Thread.State: TIMED_WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
- parking to wait for  <…> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)

"pool-73-thread-1" #358 prio=5 os_prio=0 tid=… nid=… waiting on condition […]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
at java.lang.Thread.sleep(Native Method)
at io.fabric8.kubernetes.client.utils.NonBlockingInputStreamPumper.run(NonBlockingInputStreamPumper.java:57)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
</pre>

<p>To see the problem, you can run</p>

<pre>
<span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
    podTemplate(label: BUILD_TAG, containers: [containerTemplate(name: <span class="code-quote">'ubuntu'</span>, image: <span class="code-quote">'ubuntu'</span>, command: <span class="code-quote">'sleep'</span>, args: <span class="code-quote">'infinity'</span>)]) {
node (BUILD_TAG) {
    container(<span class="code-quote">'ubuntu'</span>) {
branches = [:] <span class="code-comment">// TODO cannot use collectEntries because: java.io.NotSerializableException: groovy.lang.IntRange
</span><span class="code-keyword">for</span> (<span class="code-object">int</span> x = 0; x < 1000; x += 5) {
    def _x = x
    branches[<span class="code-quote">"sleep$x"</span>] = {
sleep time: _x, unit: <span class="code-quote">'SECONDS'</span>
sh <span class="code-quote">"""set +x; <span class="code-keyword">while</span> :; <span class="code-keyword">do</span> echo -n "</span>$_x <span class="code-quote">"; date; sleep 10; done"</span>""
    }
}
parallel branches
    }
}
    }
}
</pre>

<p>and watch via</p>

<pre>
<span class="code-object">while</span> :; <span class="code-object">do</span> jstack <span class="code-object">$pid</span> | fgrep <span class="code-quote-red">'"'</span> | sort | egrep -i <span class="code-quote-red">'ok|pool'</span> > /tmp/x; clear; cat /tmp/x; sleep 5; <span class="code-object">done</span>
</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290" target="_blank">WebSocket / OkHttp thread leak from BourneShellScript + ContainerExecDecorator</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: durable-task-plugin, kubernetes-plugin
<li><b>label(s)</b>: leak, threads
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2020-04-02T19:00:14+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The <tt>durable-task</tt> plugin runs a wrapper process which redirects the user process' stdout/err to a file and sends its exit code to another file. Thus there is no need for the agent JVM to hold onto a process handle for the wrapper; it should be fork-and-forget. In fact the <tt>Proc</tt> is discarded.</p>

<p>Unfortunately, the current implementation in <tt>BourneShellScript</tt> does not actually allow the <tt>Proc</tt> to exit until the user process also exits. On a regular agent this does not matter much. But when you run <tt>sh</tt> steps inside <tt>container</tt> on a Kubernetes agent, <tt>ContainerExecDecorator</tt> and <tt>ContainerExecProc</tt> actually keep a WebSocket open for the duration of the launched process. This consumes resources on the Kubernetes API server; it is possible to run out of connections. It also consumes three master-side Java threads per <tt>sh</tt>, like</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">"OkHttp http://…/..." #361 prio=5 os_prio=0 tid=… nid=… runnable […]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
        at java.net.SocketInputStream.read(SocketInputStream.java:171)
        at java.net.SocketInputStream.read(SocketInputStream.java:141)
        at okio.Okio$2.read(Okio.java:140)
        at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
        at okio.RealBufferedSource.request(RealBufferedSource.java:68)
        at okio.RealBufferedSource.require(RealBufferedSource.java:61)
        at okio.RealBufferedSource.readByte(RealBufferedSource.java:74)
        at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.java:117)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:101)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:274)
        at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:214)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

"OkHttp WebSocket http://…/..." #359 prio=5 os_prio=0 tid=… nid=… waiting on condition […]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  &lt;…&gt; (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
        at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
        at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
        at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

"pool-73-thread-1" #358 prio=5 os_prio=0 tid=… nid=… waiting on condition […]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(Native Method)
        at io.fabric8.kubernetes.client.utils.NonBlockingInputStreamPumper.run(NonBlockingInputStreamPumper.java:57)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
</pre>
</div></div>

<p>To see the problem, you can run</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
    podTemplate(label: BUILD_TAG, containers: [containerTemplate(name: <span class="code-quote">'ubuntu'</span>, image: <span class="code-quote">'ubuntu'</span>, command: <span class="code-quote">'sleep'</span>, args: <span class="code-quote">'infinity'</span>)]) {
        node (BUILD_TAG) {
            container(<span class="code-quote">'ubuntu'</span>) {
                branches = [:] <span class="code-comment">// TODO cannot use collectEntries because: java.io.NotSerializableException: groovy.lang.IntRange
</span>                <span class="code-keyword">for</span> (<span class="code-object">int</span> x = 0; x &lt; 1000; x += 5) {
                    def _x = x
                    branches[<span class="code-quote">"sleep$x"</span>] = {
                        sleep time: _x, unit: <span class="code-quote">'SECONDS'</span>
                        sh <span class="code-quote">"""set +x; <span class="code-keyword">while</span> :; <span class="code-keyword">do</span> echo -n "</span>$_x <span class="code-quote">"; date; sleep 10; done"</span>""
                    }
                }
                parallel branches
            }
        }
    }
}
</pre>
</div></div>

<p>and watch via</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-sh"><span class="code-object">while</span> :; <span class="code-object">do</span> jstack <span class="code-object">$pid</span> | fgrep <span class="code-quote-red">'"'</span> | sort | egrep -i <span class="code-quote-red">'ok|pool'</span> &gt; /tmp/x; clear; cat /tmp/x; sleep 5; <span class="code-object">done</span>
</pre>
</div></div></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [CloudBees-internal issue](https://cloudbees.atlassian.net/browse/CPLT2-5657)
  - [durable-task #​98](https://github.com/jenkinsci/durable-task-plugin/pull/98)
  - [kubernetes #​522](https://github.com/jenkinsci/kubernetes-plugin/pull/522)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>daniel_jeznach_img:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290?focusedId=372489&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372489">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello,
 fix for this issue causes one of our pipelines to stuck. Passing </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>-Dorg.jenkinsci.plugins.durabletask.BourneShellScript.LAUNCH_DIAGNOSTICS=true
</pre>
</div></div>
<p> parameter reverts to old behavior, so I blame running sh build step in subshell as a cause of this.</p>

<p>Pipeline code is:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent { label <span class="code-quote">"xxxxxx"</span> + (env.JOB_NAME.find(/-(u14|u16|u18|w7|w10)(?=-)/) ?:'') }
    
    options {
        timeout(time: 60, unit: <span class="code-quote">'MINUTES'</span>)
    }
        
    environment {
        TMP = <span class="code-quote">"${env.WORKSPACE}/tmp"</span>
        PROJECT_NAME = <span class="code-quote">"${env.JOB_NAME}"</span>.find(/^[^-]+/)
        PROJECT_BRANCH = <span class="code-quote">"${env.JOB_NAME}"</span>.find(/(?&lt;=^[^-]+-)[^-]+/)
        PROJECT_RELEASE = <span class="code-quote">"${env.PROJECT_BRANCH}"</span>.replace(<span class="code-quote">'MAIN'</span>, <span class="code-quote">'9999rc'</span>)
        P4_BRANCH = <span class="code-quote">"REL/${env.PROJECT_BRANCH}"</span>.replace(<span class="code-quote">'REL/MAIN'</span>, <span class="code-quote">'MAIN'</span>)
    }
    
    parameters {
        string(name: <span class="code-quote">'change'</span>, defaultValue: <span class="code-quote">'', description: '</span>Changelist to build.', trim: <span class="code-keyword">true</span>)
        string(name: <span class="code-quote">'status'</span>, defaultValue: <span class="code-quote">'', description: '</span>Changelist status (shelved/submitted).', trim: <span class="code-keyword">true</span>)
        string(name: <span class="code-quote">'review'</span>, defaultValue: <span class="code-quote">'', description: '</span>Helix Swarm review number.', trim: <span class="code-keyword">true</span>)
        string(name: <span class="code-quote">'pass'</span>, defaultValue: <span class="code-quote">'', description: '</span>Success callback URL.', trim: <span class="code-keyword">true</span>)
        string(name: <span class="code-quote">'fail'</span>, defaultValue: <span class="code-quote">'', description: '</span>Failure callback URL.', trim: <span class="code-keyword">true</span>)
    }
    
    triggers {
        pollSCM(<span class="code-quote">'@midnight'</span>)
    }
    
    stages {
        stage(<span class="code-quote">'sync'</span>) {
            steps {
                p4sync(
                    credential: <span class="code-quote">'xxxxx'</span>,
                    format: <span class="code-quote">'xxxx.xxxx-${NODE_NAME}-${JOB_NAME}-${EXECUTOR_NUMBER}_ws'</span>,
                    populate: forceClean(),
                    source: depotSource(<span class="code-quote">"<span class="code-comment">//path/${env.P4_BRANCH}/..."</span>)
</span>                )
        
                buildName <span class="code-quote">"${env.P4_CHANGELIST} ${params.status?:<span class="code-quote">''} ${params.change?:'</span>'}"</span>.trim()
                sh <span class="code-quote">'''sed '</span>s/\\\$Change: [0-9]\\+ \\\$/\$Change: <span class="code-quote">'$P4_CHANGELIST'</span> \$/<span class="code-quote">' -i xxxxx/version.py'</span>''
                    
                script {
                    def dot = env.PROJECT_RELEASE.find(/rc$/) ? <span class="code-quote">'' : '</span>.'
                    env.PACKAGE_NAME = <span class="code-quote">"${env.PROJECT_NAME}-${env.PROJECT_RELEASE}${dot}${env.P4_CHANGELIST}.tar.gz"</span>
                }
                echo <span class="code-quote">"PROJECT_NAME = $PROJECT_NAME"</span>
                echo <span class="code-quote">"PROJECT_BRANCH = $PROJECT_BRANCH"</span>
                echo <span class="code-quote">"PROJECT_RELEASE = $PROJECT_RELEASE"</span>
                echo <span class="code-quote">"P4_BRANCH = $P4_BRANCH"</span>
                echo <span class="code-quote">"PACKAGE_NAME = $PACKAGE_NAME"</span>
            }
        }
        stage(<span class="code-quote">'build'</span>) {
            steps { 
                sh '''
                #!/bin/bash -xe
        
                # Clean up possible orphans from other test sessions
                pkill -P1 -U $(id -u)  || <span class="code-keyword">true</span>
                
                mkdir $TMP
                
                virtualenv venv
                . venv/bin/activate
                
                python setup.py sdist
                pip install dist/${PROJECT_NAME}-${PROJECT_RELEASE}*.tar.gz
                xxxxxx master --print-requirements | xargs pip install
                pip install -r tests/test_requirements.txt
                '''
            }
        }
        stage(<span class="code-quote">'test'</span>) { 
            steps { 
                sh '''
                    . venv/bin/activate
                    pytest -n 4 \
                      --timeout=300 \
                      --junit-xml ${WORKSPACE}/testresults.xml \
                      --verbose \
                      --cov=${PROJECT_NAME} \
                      --cov-branch \
                      --cov-report xml:${WORKSPACE}/coverage.xml \
                      --replay-record-dir=testreplay \
                      tests || <span class="code-keyword">true</span>
                '''
                junit <span class="code-quote">'testresults.xml'</span>
                cobertura coberturaReportFile: <span class="code-quote">'coverage.xml'</span>
            }
        }
    }
}
</pre>
</div></div>

<p>It never gets to "test" stage, seems that "sh" step in "build" stage never ends.
If you need, I'll gladly provide more information.</p>
</pre>
</details>

</ul>
</details>

<p>Hello,<br/>
 fix for this issue causes one of our pipelines to stuck. Passing </p>


```
-Dorg.jenkinsci.plugins.durabletask.BourneShellScript.LAUNCH_DIAGNOSTICS=true

```
<p> parameter reverts to old behavior, so I blame running sh build step in subshell as a cause of this.</p>

<p>Pipeline code is:</p>


<pre>
pipeline {
    agent { label <span class="code-quote">"xxxxxx"</span> + (env.JOB_NAME.find(/-(u14|u16|u18|w7|w10)(?=-)/) ?:'') }
    
    options {
timeout(time: 60, unit: <span class="code-quote">'MINUTES'</span>)
    }

    environment {
TMP = <span class="code-quote">"${env.WORKSPACE}/tmp"</span>
PROJECT_NAME = <span class="code-quote">"${env.JOB_NAME}"</span>.find(/^[^-]+/)
PROJECT_BRANCH = <span class="code-quote">"${env.JOB_NAME}"</span>.find(/(?<=^[^-]+-)[^-]+/)
PROJECT_RELEASE = <span class="code-quote">"${env.PROJECT_BRANCH}"</span>.replace(<span class="code-quote">'MAIN'</span>, <span class="code-quote">'9999rc'</span>)
P4_BRANCH = <span class="code-quote">"REL/${env.PROJECT_BRANCH}"</span>.replace(<span class="code-quote">'REL/MAIN'</span>, <span class="code-quote">'MAIN'</span>)
    }
    
    parameters {
string(name: <span class="code-quote">'change'</span>, defaultValue: <span class="code-quote">'', description: '</span>Changelist to build.', trim: <span class="code-keyword">true</span>)
string(name: <span class="code-quote">'status'</span>, defaultValue: <span class="code-quote">'', description: '</span>Changelist status (shelved/submitted).', trim: <span class="code-keyword">true</span>)
string(name: <span class="code-quote">'review'</span>, defaultValue: <span class="code-quote">'', description: '</span>Helix Swarm review number.', trim: <span class="code-keyword">true</span>)
string(name: <span class="code-quote">'pass'</span>, defaultValue: <span class="code-quote">'', description: '</span>Success callback URL.', trim: <span class="code-keyword">true</span>)
string(name: <span class="code-quote">'fail'</span>, defaultValue: <span class="code-quote">'', description: '</span>Failure callback URL.', trim: <span class="code-keyword">true</span>)
    }
    
    triggers {
pollSCM(<span class="code-quote">'@​midnight'</span>)
    }
    
    stages {
stage(<span class="code-quote">'sync'</span>) {
    steps {
p4sync(
    credential: <span class="code-quote">'xxxxx'</span>,
    format: <span class="code-quote">'xxxx.xxxx-${NODE_NAME}-${JOB_NAME}-${EXECUTOR_NUMBER}_ws'</span>,
    populate: forceClean(),
    source: depotSource(<span class="code-quote">"<span class="code-comment">//path/${env.P4_BRANCH}/..."</span>)
</span>)

buildName <span class="code-quote">"${env.P4_CHANGELIST} ${params.status?:<span class="code-quote">''} ${params.change?:'</span>'}"</span>.trim()
sh <span class="code-quote">'''sed '</span>s/\\\$Change: [0-9]\\+ \\\$/\$Change: <span class="code-quote">'$P4_CHANGELIST'</span> \$/<span class="code-quote">' -i xxxxx/version.py'</span>''
    
script {
    def dot = env.PROJECT_RELEASE.find(/rc$/) ? <span class="code-quote">'' : '</span>.'
    env.PACKAGE_NAME = <span class="code-quote">"${env.PROJECT_NAME}-${env.PROJECT_RELEASE}${dot}${env.P4_CHANGELIST}.tar.gz"</span>
}
echo <span class="code-quote">"PROJECT_NAME = $PROJECT_NAME"</span>
echo <span class="code-quote">"PROJECT_BRANCH = $PROJECT_BRANCH"</span>
echo <span class="code-quote">"PROJECT_RELEASE = $PROJECT_RELEASE"</span>
echo <span class="code-quote">"P4_BRANCH = $P4_BRANCH"</span>
echo <span class="code-quote">"PACKAGE_NAME = $PACKAGE_NAME"</span>
    }
}
stage(<span class="code-quote">'build'</span>) {
    steps { 
sh '''
#!/bin/bash -xe

# Clean up possible orphans from other test sessions
pkill -P1 -U $(id -u)  || <span class="code-keyword">true</span>

mkdir $TMP

virtualenv venv
. venv/bin/activate

python setup.py sdist
pip install dist/${PROJECT_NAME}-${PROJECT_RELEASE}*.tar.gz
xxxxxx master --print-requirements | xargs pip install
pip install -r tests/test_requirements.txt
'''
    }
}
stage(<span class="code-quote">'test'</span>) { 
    steps { 
sh '''
    . venv/bin/activate
    pytest -n 4 \
      --timeout=300 \
      --junit-xml ${WORKSPACE}/testresults.xml \
      --verbose \
      --cov=${PROJECT_NAME} \
      --cov-branch \
      --cov-report xml:${WORKSPACE}/coverage.xml \
      --replay-record-dir=testreplay \
      tests || <span class="code-keyword">true</span>
'''
junit <span class="code-quote">'testresults.xml'</span>
cobertura coberturaReportFile: <span class="code-quote">'coverage.xml'</span>
    }
}
    }
}
</pre>

<p>It never gets to "test" stage, seems that "sh" step in "build" stage never ends.<br/>
If you need, I'll gladly provide more information.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290?focusedId=372514&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372514">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=daniel_jeznach_img" class="user-hover" rel="daniel_jeznach_img">daniel_jeznach_img</a> you can file a separate issue (using the <b>is caused by</b> option to <b>Link</b>) with <em>complete, minimal</em> steps to reproduce <em>from scratch</em>.</p>

<p>Offhand I suspect the <tt>pkill</tt> command is to blame, perhaps by killing too much. It is possible you did not need this to begin with—exiting a <tt>node</tt> block already sends a termination signal to all processes which inherited the (IIRC) <tt>JENKINS_MAGIC_COOKIE</tt> environment variable passed to any nested <tt>sh</tt> steps.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=daniel_jeznach_img" class="user-hover" rel="daniel_jeznach_img">daniel_jeznach_img</a> you can file a separate issue (using the <b>is caused by</b> option to <b>Link</b>) with <em>complete, minimal</em> steps to reproduce <em>from scratch</em>.</p>

<p>Offhand I suspect the <tt>pkill</tt> command is to blame, perhaps by killing too much. It is possible you did not need this to begin with—exiting a <tt>node</tt> block already sends a termination signal to all processes which inherited the (IIRC) <tt>JENKINS_MAGIC_COOKIE</tt> environment variable passed to any nested <tt>sh</tt> steps.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>daniel_jeznach_img:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290?focusedId=372557&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372557">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Yes, pkill was guilty in this case. Thank you for quick and accurate response.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Yes, pkill was guilty in this case. Thank you for quick and accurate response.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>laurent1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290?focusedId=388472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-388472">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi,</p>

<p>fix for this issue cause also a freeze during pytest execution on docker node</p>

<p>The sh command are block during execution</p>

<p>With the diagnostics parameters the execution works correctly</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">-Dorg.jenkinsci.plugins.durabletask.BourneShellScript.LAUNCH_DIAGNOSTICS=<span class="code-keyword">true</span></pre>
</div></div>
<p> </p>

<p> </p>

<p> The context is an pipeline to launch pytest with tox</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">docker_img_version=<span class="code-quote">'0.17'</span>
launch_container_with_user=<span class="code-quote">'jenkins'</span>
parser_name=<span class="code-quote">'sweep_tools'</span>
github_cred_id=<span class="code-quote">'xxxxxx'</span>
github_url=<span class="code-quote">'git@github.com:xxx.git'</span>
pipeline {
  agent {
    docker {
      image <span class="code-quote">'xxxxx/analytics/ci_process:'</span>+docker_img_version
      label <span class="code-quote">'slaves_host'</span>
      args <span class="code-quote">'-u jenkins -e POSTGRESQL_USER=xxxx -e POSTGRESQL_PASSWORD=xxxx -e POSTGRESQL_DATABASE=xxxxx'</span>
      customWorkspace <span class="code-quote">"/home/jenkins/workspace/${env.JOB_NAME}"</span>
    }
  }
  stages {
    stage(<span class="code-quote">'Delete tox cache'</span>) {
      steps {
        dir(<span class="code-quote">"${env.WORKSPACE}/.tox"</span>) {
          deleteDir()
        }
      }
    }
    stage(<span class="code-quote">'Test execution'</span>) {
      steps {
        sh <span class="code-quote">'tox'</span>
        junit <span class="code-quote">'tests/results/results.xml'</span>
        cobertura autoUpdateHealth: <span class="code-keyword">false</span>, autoUpdateStability: <span class="code-keyword">false</span>,  coberturaReportFile: <span class="code-quote">'tests/results/coverage.xml'</span>, conditionalCoverageTargets: <span class="code-quote">'70, 0, 0'</span>, failUnhealthy: <span class="code-keyword">false</span>, failUnstable: <span class="code-keyword">false</span>, lineCoverageTargets: <span class="code-quote">'80, 0, 0'</span>, maxNumberOfBuilds: 0, methodCoverageTargets: <span class="code-quote">'80, 0, 0'</span>, onlyStable: <span class="code-keyword">false</span>, sourceEncoding: <span class="code-quote">'ASCII'</span>, zoomCoverageChart: <span class="code-keyword">false</span>
      }
    }
  }
}

</pre>
</div></div>
<p> The code in tox.ini file is </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[tox]
envlist = py36[testenv]
install_command = pip install {opts} {packages}
deps =
	pytest
	pytest-cov
	pytest-flask
	chardet
commands = pytest --cov=sweep_tools --cov-report term-missing --cov-report xml:tests/results/coverage.xml --junitxml tests/results/results.xml
</pre>
</div></div>
<p> </p>

<p>Actually the command line executed in step is and freeze just after pytest execution:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">sh -c <span class="code-quote">"({ <span class="code-keyword">while</span> [ -d <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d'</span> -a \! -f <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt'</span> ]; <span class="code-keyword">do</span> touch <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-log.txt'</span>; sleep 3; done } &amp; jsc=durable-07cc39a960911d2b0363cd9d28761c7c; JENKINS_SERVER_COOKIE=$jsc <span class="code-quote">'sh'</span> -xe <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/script.sh'</span> &gt; <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-log.txt'</span> 2&gt;&amp;1; echo $? &gt; <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span>; mv <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span> <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt'</span>; wait) &gt;&amp;- 2&gt;&amp;- &amp;"</span></pre>
</div></div>
<p> </p>

<p>I can fix the the freeze with following command line:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">sh -c <span class="code-quote">"({ <span class="code-keyword">while</span> [ -d <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d'</span> -a \! -f <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt'</span> ]; <span class="code-keyword">do</span> touch <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-log.txt'</span>; sleep 3; done } &amp; jsc=durable-07cc39a960911d2b0363cd9d28761c7c; JENKINS_SERVER_COOKIE=$jsc <span class="code-quote">'sh'</span> -xe <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/script.sh'</span> &gt; <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-log.txt'</span> 2&gt;&amp;1; echo $? &gt; <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span>; mv <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span> <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@tmp/durable-d4d01f8d/jenkins-result.txt'</span>; wait) &gt;&amp;- 2&gt;&amp;-"</span></pre>
</div></div>
<p> </p>

<p>In the code the line impacted is BourneShellScript.java</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">cmd.addAll(Arrays.asList(<span class="code-quote">"sh"</span>, <span class="code-quote">"-c"</span>, <span class="code-quote">"("</span> + cmdString + <span class="code-quote">") &gt;&amp;- 2&gt;&amp;- &amp;"</span>));</pre>
</div></div>
<p> </p>

<p>It can be replaced by</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">cmd.addAll(Arrays.asList(<span class="code-quote">"sh"</span>, <span class="code-quote">"-c"</span>, <span class="code-quote">"("</span> + cmdString + <span class="code-quote">") &gt;&amp;- 2&gt;&amp;-"</span>));</pre>
</div></div>
<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi,</p>

<p>fix for this issue cause also a freeze during pytest execution on docker node</p>

<p>The sh command are block during execution</p>

<p>With the diagnostics parameters the execution works correctly</p>

<pre>
-Dorg.jenkinsci.plugins.durabletask.BourneShellScript.LAUNCH_DIAGNOSTICS=<span class="code-keyword">true</span></pre>
<p> </p>

<p> </p>

<p> The context is an pipeline to launch pytest with tox</p>

<pre>
docker_img_version=<span class="code-quote">'0.17'</span>
launch_container_with_user=<span class="code-quote">'jenkins'</span>
parser_name=<span class="code-quote">'sweep_tools'</span>
github_cred_id=<span class="code-quote">'xxxxxx'</span>
github_url=<span class="code-quote">'git@​github.com:xxx.git'</span>
pipeline {
  agent {
    docker {
      image <span class="code-quote">'xxxxx/analytics/ci_process:'</span>+docker_img_version
      label <span class="code-quote">'slaves_host'</span>
      args <span class="code-quote">'-u jenkins -e POSTGRESQL_USER=xxxx -e POSTGRESQL_PASSWORD=xxxx -e POSTGRESQL_DATABASE=xxxxx'</span>
      customWorkspace <span class="code-quote">"/home/jenkins/workspace/${env.JOB_NAME}"</span>
    }
  }
  stages {
    stage(<span class="code-quote">'Delete tox cache'</span>) {
      steps {
dir(<span class="code-quote">"${env.WORKSPACE}/.tox"</span>) {
  deleteDir()
}
      }
    }
    stage(<span class="code-quote">'Test execution'</span>) {
      steps {
sh <span class="code-quote">'tox'</span>
junit <span class="code-quote">'tests/results/results.xml'</span>
cobertura autoUpdateHealth: <span class="code-keyword">false</span>, autoUpdateStability: <span class="code-keyword">false</span>,  coberturaReportFile: <span class="code-quote">'tests/results/coverage.xml'</span>, conditionalCoverageTargets: <span class="code-quote">'70, 0, 0'</span>, failUnhealthy: <span class="code-keyword">false</span>, failUnstable: <span class="code-keyword">false</span>, lineCoverageTargets: <span class="code-quote">'80, 0, 0'</span>, maxNumberOfBuilds: 0, methodCoverageTargets: <span class="code-quote">'80, 0, 0'</span>, onlyStable: <span class="code-keyword">false</span>, sourceEncoding: <span class="code-quote">'ASCII'</span>, zoomCoverageChart: <span class="code-keyword">false</span>
      }
    }
  }
}

</pre>
<p> The code in tox.ini file is </p>

<pre>
[tox]
envlist = py36[testenv]
install_command = pip install {opts} {packages}
deps =
	pytest
	pytest-cov
	pytest-flask
	chardet
commands = pytest --cov=sweep_tools --cov-report term-missing --cov-report xml:tests/results/coverage.xml --junitxml tests/results/results.xml
</pre>
<p> </p>

<p>Actually the command line executed in step is and freeze just after pytest execution:</p>

<pre>
sh -c <span class="code-quote">"({ <span class="code-keyword">while</span> [ -d <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d'</span> -a \! -f <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt'</span> ]; <span class="code-keyword">do</span> touch <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-log.txt'</span>; sleep 3; done } & jsc=durable-07cc39a960911d2b0363cd9d28761c7c; JENKINS_SERVER_COOKIE=$jsc <span class="code-quote">'sh'</span> -xe <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/script.sh'</span> > <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-log.txt'</span> 2>&1; echo $? > <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span>; mv <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span> <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt'</span>; wait) >&- 2>&- &"</span></pre>
<p> </p>

<p>I can fix the the freeze with following command line:</p>

<pre>
sh -c <span class="code-quote">"({ <span class="code-keyword">while</span> [ -d <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d'</span> -a \! -f <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt'</span> ]; <span class="code-keyword">do</span> touch <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-log.txt'</span>; sleep 3; done } & jsc=durable-07cc39a960911d2b0363cd9d28761c7c; JENKINS_SERVER_COOKIE=$jsc <span class="code-quote">'sh'</span> -xe <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/script.sh'</span> > <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-log.txt'</span> 2>&1; echo $? > <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span>; mv <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt.tmp'</span> <span class="code-quote">'/home/jenkins/workspace/auto_ci/sweep_tools/dev@​tmp/durable-d4d01f8d/jenkins-result.txt'</span>; wait) >&- 2>&-"</span></pre>
<p> </p>

<p>In the code the line impacted is BourneShellScript.java</p>

<pre>
cmd.addAll(Arrays.asList(<span class="code-quote">"sh"</span>, <span class="code-quote">"-c"</span>, <span class="code-quote">"("</span> + cmdString + <span class="code-quote">") >&- 2>&- &"</span>));</pre>
<p> </p>

<p>It can be replaced by</p>

<pre>
cmd.addAll(Arrays.asList(<span class="code-quote">"sh"</span>, <span class="code-quote">"-c"</span>, <span class="code-quote">"("</span> + cmdString + <span class="code-quote">") >&- 2>&-"</span>));</pre>
<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-58290?focusedId=388481&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-388481">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=laurent1" class="user-hover" rel="laurent1">laurent1</a> please do not reopen. If you believe you have discovered a regression, file a separate issue linked to this one with complete, minimal steps to reproduce from scratch.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=laurent1" class="user-hover" rel="laurent1">laurent1</a> please do not reopen. If you believe you have discovered a regression, file a separate issue linked to this one with complete, minimal steps to reproduce from scratch.</p>
